### PR TITLE
[PREVIEW] Solicitor journey branch

### PR DESCRIPTION
**Pull request by tchow8**
All the commits required for the Solicitor's journey will be here.

This includes the following:

Gender enum bug fix for solicitor.
PBA callbacks. Getting current fee details and also to process PBA payments if PBA is the chosen payment type.
Other code cleanup.

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/divorce/errorhandler/CcdCallbackExceptionHandler.java [hunk 1]
```diff
@@ -46,6 +48,24 @@
...
+    }
+
+    @ExceptionHandler(PaymentFailedException.class)
+    public ResponseEntity<CCDCallbackResponse> handleFailedPaymentException(
+        PaymentFailedException exception,
+        HttpServletRequest request) {
+
+        String customMessage = "Payment has failed, please try again.";
+        return generateBadRequestResponse(customMessage, request);
+    }
```
Comment (david-gimelle) [inline src/main/java/uk/gov/hmcts/reform/divorce/errorhandler/CcdCallbackExceptionHandler.java]: why not log the exception?

File: src/main/java/uk/gov/hmcts/reform/divorce/pay/models/request/FeesItem.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    @JsonProperty("code")
+    private String code;
+
+    @JsonProperty("calculated_amount")
+    private String calculatedAmount;
+
+    @JsonProperty("version")
+    private String version;
+
+    public void setCalculatedAmount(String calculatedAmount) {
```
Comment (david-gimelle) [inline src/main/java/uk/gov/hmcts/reform/divorce/pay/models/request/FeesItem.java]: do we have any unit test for this function?

File: src/main/java/uk/gov/hmcts/reform/divorce/errorhandler/CcdCallbackExceptionHandler.java (hunk 1 above)
Comment (tchow8) [reply src/main/java/uk/gov/hmcts/reform/divorce/errorhandler/CcdCallbackExceptionHandler.java]: Azure should automatically capture this exception so e don't need to specifically log it I believe.

File: src/test/java/uk/gov/hmcts/reform/divorce/pay/PaymentServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import uk.gov.hmcts.reform.divorce.transformservice.domain.model.ccd.CoreCaseData;
+import uk.gov.hmcts.reform.divorce.transformservice.domain.model.ccd.FeesItem;
+import uk.gov.hmcts.reform.divorce.transformservice.domain.model.ccd.OrderSummary;
+
+import java.util.ArrayList;
+
+import static org.mockito.Mockito.mock;
+
+@RunWith(JUnit4.class)
+public class PaymentServiceTest {
```
Comment (mthava) [inline src/test/java/uk/gov/hmcts/reform/divorce/pay/PaymentServiceTest.java]: missing quite a few tests scenarios

File: src/test/java/uk/gov/hmcts/reform/divorce/transformservice/controller/CcdCallbackControllerTest.java [hunk 4]
```diff
@@ -207,6 +224,156 @@ public void givenNoPetitionerEmail_whenPetitionSubmittedEndpointIsCalled_thenEma
...
+        coreCaseData.setOrderSummary(orderSummary);
+        caseDetails.setCaseData(coreCaseData);
+        submittedCase.setCaseDetails(caseDetails);
+
+        when(feesAndPaymentService.getPetitionIssueFee()).thenReturn(Fee.builder()
+            .feeCode("2").amount(555.00).version(2).build());
+        mvc.perform(post(PETITION_ISSUE_FEES_URL)
+            .content(ObjectMapperTestUtil.convertObjectToJsonString(submittedCase))
+            .contentType(MediaType.APPLICATION_JSON_UTF8));
+        verify(feesAndPaymentService, times(1)).getPetitionIssueFee();
```
Comment (mthava) [inline src/test/java/uk/gov/hmcts/reform/divorce/transformservice/controller/CcdCallbackControllerTest.java]: minor add space between GWT for better readability

File: src/test/java/uk/gov/hmcts/reform/divorce/transformservice/controller/CcdCallbackControllerTest.java [hunk 5]
```diff
@@ -207,6 +224,156 @@ public void givenNoPetitionerEmail_whenPetitionSubmittedEndpointIsCalled_thenEma
...
+
+    }
+
+    @Test
+    public void givenFeesNotAvailable_whenSolicitorFeesIsCalled_thenExpectException() throws Exception {
```
Comment (mthava) [inline src/test/java/uk/gov/hmcts/reform/divorce/transformservice/controller/CcdCallbackControllerTest.java]: minor add space between GWT for better readability

File: src/test/java/uk/gov/hmcts/reform/divorce/transformservice/controller/CcdCallbackControllerTest.java [hunk 6]
```diff
@@ -207,6 +224,156 @@ public void givenNoPetitionerEmail_whenPetitionSubmittedEndpointIsCalled_thenEma
...
+        mvc.perform(post(PETITION_ISSUE_FEES_URL)
+            .content(ObjectMapperTestUtil.convertObjectToJsonString(submittedCase))
+            .contentType(MediaType.APPLICATION_JSON_UTF8)).andExpect(
+            jsonPath("$.errors" , Matchers.hasSize(1)));
+        verify(feesAndPaymentService, times(1)).getPetitionIssueFee();
+
+    }
+
+    @Test
+    public void givenCallbackReceived_whenNonPBAPaymentMethodIsUsed_thenExpectToSkipPBAPayment() throws Exception {
```
Comment (mthava) [inline src/test/java/uk/gov/hmcts/reform/divorce/transformservice/controller/CcdCallbackControllerTest.java]: minor add space between GWT for better readability

File: src/test/java/uk/gov/hmcts/reform/divorce/pay/PaymentServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+    @Before
+    public void setup() {
+        this.paymentService =  mock(PaymentService.class);
+        this.paymentService = new PaymentService(paymentServiceClient, mock(AuthTokenGenerator.class));
+    }
+
+    @Test
+    public void assertValidPayment() {
+        Assert.assertTrue(true);
```
[the remark this answers is not in the document]
Comment (tchow8) [reply src/test/java/uk/gov/hmcts/reform/divorce/pay/PaymentServiceTest.java]: Haha I just saw this as well. Looks like there needs to be some cleanup. I'll raise a ticket for these.

**Comment by tchow8** [comment]
Can I request approval for now and we can fix all the issues through:
https://tools.hmcts.net/jira/browse/DIV-3094

Based off manual testing was ongoing for more than a few days.

**Comment by tchow8** [comment]
Not passing right now as config in AAT has been changed. It requires solicitors config in order to pass the functional test.

**Comment by nsingamsetty** [comment]
Tested and working fine as Expected. Merging the PR.

**Pull request merged**